horovod/common/control_manager.h

```diff
@@ -0,0 +1,102 @@
+//
+// Created by Sihan Zeng on 2019-06-07.
+//
+
+#ifndef HOROVOD_CONTROL_MANAGER_H
+#define HOROVOD_CONTROL_MANAGER_H
+
+#include <iostream>
+#include <vector>
+#include "half.h"
+#include "mpi_context.h"
+
+#if HAVE_GLOO
+#include "gloo_context.h"
+#endif
+
+#define IN_PLACE MPI_IN_PLACE
+
+namespace horovod {
+namespace common {
+
+enum OpType { HOROVOD_SUM = 0, HOROVOD_BAND = 1, HOROVOD_BOR = 2 };
+
+class Controller {
+public:
+  enum ControllerType { MPI = 0, GLOO = 1 };
+
+  // Functions must be overridden by concrete controller
+  virtual void Initialize() = 0;
+  virtual void Finalize() = 0;
+
+  virtual int GetTypeSize(DataType dtype) = 0;
+
+  virtual void AllReduce(const void* sendbuf, void* recvbuf, int count,
+                         DataType datatype, OpType optype,
+                         Communicator comm) = 0;
+  virtual void Gather(const void* sendbuf, int sendcount, DataType sendtype,
+                      void* recvbuf, int recvcount, DataType recvtype, int root,
+                      Communicator comm) = 0;
+  virtual void AllGather(const void* sendbuf, int sendcount, DataType sendtype,
+                         void* recvbuf, int recvcount, DataType recvtype,
+                         Communicator comm) = 0;
+  virtual void Gatherv(const void* sendbuf, int sendcount, DataType sendtype,
+                       void* recvbuf, const int recvcount[], const int displs[],
+                       DataType recvtype, int root, Communicator comm) = 0;
+  virtual void Bcast(void* buffer, int count, DataType datatype, int root,
+                     Communicator comm) = 0;
+  virtual void Barrier(Communicator comm) = 0;
+
+#if HAVE_GLOO
+  virtual void InitializeGlooCtx(GlooContext &gloo_ctx, const char* iface) = 0;
+#endif
+
+  // general controller functions
+  void SetRank(const int* ranks, int nrank);
```

Comment:
I'd go ahead and put the implementations in the header too to clean things up.

Reply:
Done.